### PR TITLE
Keep queries in the queue when task count is high

### PR DESCRIPTION
**Pull request by viczhang861**
Use total running task count as a proxy for cluster load and keep 
queries in the queue if load is too high.

- Attached figure shows the result from `getTaskLimitExceeded()` in a real cluster,  correlation of this chart with queue size makes debugging easier.

Screen Shot 2020-03-04 at 7 41 38 PM

```
== RELEASE NOTES ==

General Changes
* Add support for stopping new query execution when total number of tasks exceeds a set limit to help with reliability. The limit can be set by the cluster property by ``experimental.max-total-running-task-count-to-halt-scheduling``.
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 1]
```diff
@@ -258,6 +259,19 @@ public int getMaxQueryRunningTaskCount()
...
 
+    @Config("enqueue-query-when-task-limit-exceeded")
+    @ConfigDescription("Keep queries in queue if max-total-running-task-count is violated")
+    public QueryManagerConfig setEnqueueQueryWhenTaskLimitExceeded(boolean enqueueQueryWhenTaskLimitExceeded)
+    {
+        this.enqueueQueryWhenTaskLimitExceeded = enqueueQueryWhenTaskLimitExceeded;
+        return this;
+    }
+
+    public boolean getEnqueueQueryWhenTaskLimitExceeded()
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: get -> is?

**Comment by mbasmanova** [comment]
@viczhang861 Perhaps, clarify the RN:

```
* Keep queries in queue when total number of running tasks exceeds a threshold specified by XXX configuration property.
```

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 2]
```diff
@@ -68,20 +70,26 @@
...
+    private final ResourceGroupManager<?> resourceGroupManager;
     private final ScheduledExecutorService queryManagementExecutor;
 
     @GuardedBy("this")
     private ScheduledFuture<?> backgroundTask;
 
-    public QueryTracker(QueryManagerConfig queryManagerConfig, ScheduledExecutorService queryManagementExecutor)
+    public QueryTracker(
+            QueryManagerConfig queryManagerConfig,
+            ResourceGroupManager resourceGroupManager,
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: This is going to be really tough to fix once we move to a world where resource manager is out of process.  I would prefer that we think of a way of accomplishing this without having to set an atomic boolean on the resource group manager itself.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 3]
```diff
@@ -258,8 +258,8 @@ public int getMaxQueryRunningTaskCount()
         return maxQueryRunningTaskCount;
     }
 
-    @Config("experimental.max-query-running-task-count")
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: This very PR is changing the semantics of this field.  Isn't it premature to rename this?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 4]
```diff
@@ -862,6 +863,10 @@ private boolean canRunMore()
                 return false;
             }
 
+            if (((RootInternalResourceGroup) root).isResourceLimitExceeded()) {
+                return false;
+            }
+
```
Comment (mayankgarg1990) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: I don't think this is the correct semantic. If `isResourceLimitExceeded` is expected to be a `root` level property, then we should just prevent running `internalStartNext` from `RootInternalResourceGroup.processQueuedQueries`. As of now I feel there is too much plumbing done through the resource groups which can be completely avoided.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 5]
```diff
@@ -935,5 +942,16 @@ public synchronized void generateCpuQuota(long elapsedSeconds)
...
+        public void setTaskLimitExceeded(boolean exceeded)
+        {
+            this.taskLimitExceeded.set(exceeded);
+        }
+
+        private boolean isResourceLimitExceeded()
+        {
+            // can be expanded to include other metrics
+            return taskLimitExceeded.get();
+        }
```
Comment (mayankgarg1990) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: If we follow my suggestion above, we can avoid this call. I will say that the plumbing is not worth it and we should just implement it when we have a concrete use case.

**Comment by mayankgarg1990** [comment]
Going by the release note guidelines - https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#phrasing

```
* Add support for halting scheduling when the number of tasks exceeds a set limit to help with coordinator reliability. The limit can be set by the cluster property ``xxx``.
```

Also, for the PR title, can we be clear about `number of tasks` instead of using `cluster load`?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java (hunk 3 above)
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: @mayankgarg1990  @tdcmeehan   I changed the new config introduced in this PR to `experimental.halt-query-scheduling-when-task-limit-exceeded` and added `@LegacyConfig()` for old ones, which were introduced in v0.225 and would continue to exist.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 6]
```diff
@@ -235,12 +245,19 @@ private void enforceTaskLimits()
...
             if (runningTaskCount > highestRunningTaskCount) {
                 highestRunningTaskCount = runningTaskCount;
                 highestRunningTaskQuery = Optional.of(query);
             }
         }
 
+        if (enqueueQueryWhenTaskLimitExceeded && totalRunningTaskCount <= maxTotalRunningTaskCount) {
+            resourceGroupManager.setTaskLimitExceeded(false);
+        }
+
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: Thanks,  good point,  early decision is not needed here, this for loop is fast enough.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 2 above)
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: @tdcmeehan Eventually, we should use CPU/Memory/IO as resource budget inside each resource group (currently it is simply query count?).  Before that happens, task count is a proxy to fix deficiency in resource group manager. A global task count limit prevents scheduling a burst of new queries when cluster load is high, similar to total query count limit but more powerful.    The added boolean value `taskLimitExceeded` is light weighted and could be easily deprecated whenever there is better solution.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: I understand that the flag is lightweight, and it makes sense that we should keep it as a flag to avoid having to deprecate the field once we move to smarter resource placement.  That being said, there doesn't seem to be a good reason to introduce this dependency in `QueryTracker`.  Resource groups already keep track of all running queries and have the full information of the running task count, we just need to access it.  Why not make the decision on whether or not to globally prevent dispatching in `QueryTracker` when we can keep it local to resource groups?  It just seems to confusing to put some of that logic here.

Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: I got your point.
-  Traversing resource group tree while taking lock to calculate total number of tasks is more expensive than doing the same thing inside `QueryTracker`. It is duplicated calculation since `QueryTracker` already has that information.
-  The dependency is already there between `SqlQueryManager` and `ResourceGroupManager`,  what you suggested is to migrate some functionalities of `SqlQueryManager` to `ResourceGroupManager`, which is nice.  Given this is how currently it is designed, I still think `QueryTracker` is the best fit and clean implementation.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 7]
```diff
@@ -51,6 +51,7 @@
     private int stageCountWarningThreshold = 50;
     private int maxTotalRunningTaskCount = Integer.MAX_VALUE;
     private int maxQueryRunningTaskCount = Integer.MAX_VALUE;
+    private int maxTotalRunningTaskCountToHaltScheduling = Integer.MAX_VALUE;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Maybe it makes sense to make this optional, so we don't need a magic number (MAX_VALUE).

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Alternatively, a negative number would never make sense here, so instead of MAX_VALUE we could simply check if it's non-negative.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java [hunk 8]
```diff
@@ -278,6 +295,46 @@ private static int getQueriesQueuedOnInternal(InternalResourceGroup resourceGrou
...
+        int taskCount = 0;
+        for (RootInternalResourceGroup rootGroup : rootGroups) {
+            synchronized (rootGroup) {
+                taskCount += calculateRunningTaskCount(rootGroup);
+            }
+        }
+        return taskCount;
+    }
+
+    private int calculateRunningTaskCount(InternalResourceGroup resourceGroup)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java]: This might conflict with what @mayankgarg1990 was suggesting, but for consistency with how we compute other stats I think we could move this into `InternalResourceGroup`

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQueryTaskLimit.java [hunk 9]
```diff
@@ -26,42 +28,55 @@
...
+                .setSchema("sf1000")
+                .build();
     }
 
     @AfterClass(alwaysRun = true)
     public void shutdown()
+            throws Exception
     {
         executor.shutdownNow();
+        assertTrue(executor.awaitTermination(10, SECONDS));
```
Comment (tdcmeehan) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQueryTaskLimit.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java (hunk 8 above)
Comment (mayankgarg1990) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java]: If we are moving num running tasks to be a stat like `cpuUsageMillis` then yes, lets move it in to the `InternalResourceGroup` class. If we want to have some kind of Supplier which indicates to pause scheduling to `InternalResourceGroupManager` then lets just operate at the root level. So I am fine either ways - just not with mixing them together

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java (hunk 7 above)
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: We use MAX_VALUE for some other defaulted values, e.g. `maxDriversPerTask` :)

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java (hunk 8 above)
Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java]: Moved to `InternalResourceGroup`

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQueryTaskLimit.java (hunk 9 above)
Comment (viczhang861) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestQueryTaskLimit.java]: Not directly related to queueing test,  I created a separate PR to avoid confusion

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 10]
```diff
@@ -244,7 +244,8 @@ public int getMaxTotalRunningTaskCount()
         return maxTotalRunningTaskCount;
     }
 
-    @Config("experimental.max-total-running-task-count")
+    @Config("max-total-running-task-count-to-kill-query")
+    @LegacyConfig("experimental.max-total-running-task-count")
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: Isn't the idea of "experimental" configs is that they can change without notice and therefore don't require @LegacyConfig-style transition?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 11]
```diff
@@ -259,6 +260,19 @@ public int getMaxQueryRunningTaskCount()
         return maxQueryRunningTaskCount;
     }
 
+    @Config("experimental.max-total-running-task-count-to-halt-scheduling")
+    @ConfigDescription("Keep queries in the queue if total task count exceeds experimental.max-total-running-task-count-to-halt-scheduling")
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: nit: I suggest to not repeat the property name in the description: `Keep queries in the queue if total task count exceeds this threshold`

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 12]
```diff
@@ -581,6 +583,23 @@ public InternalResourceGroup getOrCreateSubGroup(String name)
         }
     }
 
+    public int getRunningTaskCount()
+    {
+        if (subGroups().isEmpty()) {
+            return runningQueries.stream()
+                    .filter(query -> query instanceof SqlQueryExecution)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: nit: SqlQueryExecution.class::isInstance

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 13]
```diff
@@ -935,5 +960,15 @@ public synchronized void generateCpuQuota(long elapsedSeconds)
                 internalGenerateCpuQuota(elapsedSeconds);
             }
         }
+
+        public void setTaskLimitExceeded(boolean exceeded)
+        {
+            this.taskLimitExceeded.set(exceeded);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: nit: drop "this."

**Comment by wenleix** [comment]
@viczhang861 : Would you mind merge this after https://github.com/prestodb/presto/pull/14104 is merged? Thanks!

**Comment by viczhang861** [comment]
> @viczhang861 : Would you mind merge this after #14104 is merged? Thanks!

Ok,  cc @mbasmanova

File: presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java [hunk 14]
```diff
@@ -258,6 +259,19 @@ public int getMaxQueryRunningTaskCount()
         return maxQueryRunningTaskCount;
     }
 
+    @Config("experimental.max-total-running-task-count-to-halt-scheduling")
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: nit: `scheduling` usually means task scheduling or stage scheduling in Presto. What about 
```
experimental.max-total-running-task-count-for-new-query-execution
```

Comment (viczhang861) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java]: > experimental.max-total-running-task-count-for-new-query-execution

Changed to ``experimental.max-total-running-task-count-to-not-execute-new-query``

**Comment by wenleix** [comment]
Travis failure is irrelevant :

```
[ERROR] testTaskUpdateSizeLimit(com.facebook.presto.tests.TestTpchTaskUpdateSizeLimitIntegrationTest)  Time elapsed: 0.227 s  <<< FAILURE!
java.lang.AssertionError: Expected exception message 'Could not communicate with the remote task. The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes. (127.0.0.1:34124)' to match 'TaskUpdate size of [0-9]* Bytes has exceeded the limit of 100 Bytes' for query: SELECT * FROM ORDERS LIMIT 10
	at org.testng.Assert.fail(Assert.java:83)
	at com.facebook.presto.tests.QueryAssertions.assertExceptionMessage(QueryAssertions.java:323)
	at com.facebook.presto.tests.QueryAssertions.assertQueryFails(QueryAssertions.java:304)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQueryFails(AbstractTestQueryFramework.java:235)
	at com.facebook.presto.tests.TestTpchTaskUpdateSizeLimitIntegrationTest.testTaskUpdateSizeLimit(TestTpchTaskUpdateSizeLimitIntegrationTest.java:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Could not communicate with the remote task. The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes. (127.0.0.1:34124)
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:115)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:398)
	at com.facebook.presto.tests.QueryAssertions.assertQueryFails(QueryAssertions.java:300)
	... 15 more
Caused by: com.facebook.presto.spi.PrestoException: Could not communicate with the remote task. The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes. (127.0.0.1:34124)
	at com.facebook.presto.server.remotetask.ContinuousTaskStatusFetcher.updateTaskStatus(ContinuousTaskStatusFetcher.java:251)
	at com.facebook.presto.server.remotetask.HttpRemoteTask.abort(HttpRemoteTask.java:774)
	at com.facebook.presto.server.remotetask.HttpRemoteTask.failTask(HttpRemoteTask.java:887)
	at com.facebook.presto.server.remotetask.HttpRemoteTask.sendUpdate(HttpRemoteTask.java:649)
	at com.facebook.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:78)
	... 3 more
```

**Comment by wenleix** [comment]
Merged #14194, thanks for the contribution!
